ChemMaster+

sleepyyapril opened the pull request:
<!-- If you are new to the Delta-V repository, please read the [Contributing Guidelines](https://github.com/DeltaV-Station/Delta-v/blob/master/CONTRIBUTING.md) -->

## About the PR
<!-- What did you change? -->

Changes how the ChemMaster works:
1. Removes the amount buttons and instead uses a textbox that resets whenever a value is entered or focus ends.
2. Shrinks the ChemMaster again.
3. Adds sorting options, including for quantity, last added, and a sort option for the (default) alphabetical order.
4. Sorting options save via the ChemMaster itself, not per-user.
5. Adds a pill buffer for all my pill-makers!

## Why / Balance
<!-- Discuss how this would affect game balance or explain why it was changed. Link any relevant discussions or issues. -->
Sorting the ChemMaster by default is bad, let people have the right to choose how it sorts.
Also lets you set your own quantity to transfer/discard.

## Technical details
<!-- Summary of code changes for easier review. -->
1. Adds TransferringAmount and SortMethod to ChemMasterComponent, to save separately for each ChemMaster.
2. Edits ChemMasterUiState to support the above.
3. Shrinks the ChemMaster's UI because we no longer need that much width.

Mostly UI changes.

## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc).
Small fixes/refactors are exempt. -->

Video showcase:
https://ptb.discord.com/channels/968983104247185448/1206353544186171482/1333324287188402277

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [x] I have tested all added content and changes.
- [x] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them. -->

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelogs must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
<!--
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->
:cl:
- add: Added ChemMaster sorting and a custom amount input.
- add: Added a ChemMaster pill buffer.

sleepyyapril committed:
ChemMaster+ (#1585)

<!--
This is a semi-strict format, you can add/remove sections as needed but
the order/format should be kept the same
Remove these comments before submitting
-->

<!--
Explain this PR in as much detail as applicable

Some example prompts to consider:
How might this affect the game? The codebase?
What might be some alternatives to this?
How/Who does this benefit/hurt [the game/codebase]?
-->

Changes how the ChemMaster works:
1. Removes the amount buttons and instead uses a textbox that resets
whenever a value is entered or focus ends.
2. Shrinks the ChemMaster again.
3. Adds sorting options, including for quantity, last added, and a sort
option for the (default) alphabetical order.
4. Sorting options save via the ChemMaster itself, not per-user.

Video showcase:
https://discord.com/channels/1218698320155906090/1218698321053356060/1330129166384894046

---

<!--
You can add an author after the `:cl:` to change the name that appears
in the changelog (ex: `:cl: Death`)
Leaving it blank will default to your GitHub display name
This includes all available types for the changelog
-->

:cl:
- add: Added sorting options to the ChemMaster.
- add: Added the ability to input custom amounts into the ChemMaster via
a textbox that resets on change.
- tweak: The width of the ChemMaster UI has been lowered.
- remove: Removed quantity buttons from the ChemMaster.

---------

Co-authored-by: VMSolidus <evilexecutive@gmail.com>

sleepyyapril commented:
I could theoretically upstream.

sleepyyapril committed:
Fix for not having DiscardReagents

ThataKat commented:
Direction is reviewing, get back to you in about 24 hours!

rosieposieeee commented:
might be a PITA, but this exclusively touches upstream files, so i would suggest opening a parallel PR and seeing if that gets accepted

sleepyyapril commented:
> might be a PITA, but this exclusively touches upstream files, so i would suggest opening a parallel PR and seeing if that gets accepted

Not able to until one of the ChemMaster PR snipe changes gets merged.

sleepyyapril commented:
I'm going over this, and I'm likely going to redo the entire ChemMaster systems as a whole. You can merge this as-is, though.

dvir001 commented:
> I'm going over this, and I'm likely going to redo the entire ChemMaster systems as a whole. You can merge this as-is, though.

No comments, no merge.

ThataKat commented:
Seen closed. Direction recommends trying this type of change upstream first anyhow.

sleepyyapril commented:
> Seen closed. Direction recommends trying this type of change upstream first anyhow.

Again, as stated, I can't. Will certainly try when that other PR is merged on WizDen's side.

sleepyyapril committed:
ChemMaster Improvements (#1650)

Hate having to use the same buffer for pills? Want sorting and container
solutions for output? No need to wait _any_ longer! A brand new pill
buffer, a sort feature _for_ the new pill buffer **and** you have your
amount buttons back!

Showcase:
https://ptb.discord.com/channels/1218698320155906090/1218698321053356060/1332224976803074123

:cl:
- add: A new ChemMaster experience has been granted to the people of
Einstein Engines. Includes a pill buffer!

sleepyyapril committed:
Fix Chemmaster Two (#1667)

simplify this shitcode

i borked it last time, sorry

---------

Signed-off-by: sleepyyapril <123355664+sleepyyapril@users.noreply.github.com>
Co-authored-by: VMSolidus <evilexecutive@gmail.com>

sleepyyapril committed:
Merge branch 'master' of https://github.com/DeltaV-Station/Delta-v into port-chemmaster-plus

sleepyyapril committed:
Fixed network fuckery

sleepyyapril committed:
Merge branch 'master' of https://github.com/DeltaV-Station/Delta-v into port-chemmaster-plus

sleepyyapril committed:
Use separate counts for the separate buffers.

sleepyyapril commented:
Re-opened with new video showcase

sleepyyapril committed:
You want comments? I'll give you comments.

sleepyyapril committed:
Remove unnecessary addition

sleepyyapril commented:
How many comments is too many?

deltanedas commented on Content.Client/Chemistry/UI/ChemMasterBoundUserInterface.cs:
```diff
@@ -10,15 +10,11 @@ namespace Content.Client.Chemistry.UI
 
-        public ChemMasterBoundUserInterface(EntityUid owner, Enum uiKey) : base(owner, uiKey)
-        {
-        }
```
this just creates conflicts 4 no raisin, undo this

deltanedas commented on Content.Client/Chemistry/UI/ChemMasterWindow.xaml:
comment xaml changes too its just xml `<!-- -->`

deltanedas commented on Content.Server/Chemistry/Components/ChemMasterComponent.cs:
```diff
@@ -20,5 +20,11 @@ public sealed partial class ChemMasterComponent : Component
+        public int SortMethod;
+
+        [DataField]
+        public int TransferringAmount;
```
```cs
/// <summary>
/// DeltaV: ...
/// </summary>
```

document them

deltanedas commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -38,10 +38,14 @@ public sealed class ChemMasterSystem : EntitySystem
         [Dependency] private readonly StorageSystem _storageSystem = default!;
         [Dependency] private readonly LabelSystem _labelSystem = default!;
         [Dependency] private readonly ISharedAdminLogManager _adminLogger = default!;
+        [Dependency] private readonly IEntityManager _entityManager = default!;
```
remove

deltanedas commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -38,10 +38,14 @@ public sealed class ChemMasterSystem : EntitySystem
         private const string PillPrototypeId = "Pill";
 
+        [ValidatePrototypeId<EntityPrototype>]
+        private const string PillCanisterPrototypeId = "PillCanister";
```
EntProtoId instead of the annotation + comment

deltanedas commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -56,27 +60,41 @@ public override void Initialize()
             SubscribeLocalEvent<ChemMasterComponent, ChemMasterCreatePillsMessage>(OnCreatePillsMessage);
             SubscribeLocalEvent<ChemMasterComponent, ChemMasterOutputToBottleMessage>(OnOutputToBottleMessage);
+            SubscribeLocalEvent<ChemMasterComponent, ChemMasterSortMethodUpdated>(OnSortMethodUpdated);
+            SubscribeLocalEvent<ChemMasterComponent, ChemMasterTransferringAmountUpdated>(OnTransferringAmountUpdated);
```
comments

deltanedas commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -56,27 +60,41 @@ public override void Initialize()
-        {
+        private void SubscribeUpdateUiState<T>(Entity<ChemMasterComponent> ent, ref T ev) =>
             UpdateUiState(ent);
-        }
```
untouch

deltanedas commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -56,27 +60,41 @@ public override void Initialize()
+                chemMaster.PillDosageLimit,
+                updateLabel,
+                ent.Comp.SortMethod,
+                ent.Comp.TransferringAmount);
```
dont reformat this, comment on the 2 actually new fields and removal of the output container part

deltanedas commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -56,27 +60,41 @@ public override void Initialize()
+            if (!_solutionContainerSystem.TryGetSolution(owner, SharedChemMaster.PillBufferSolutionName, out _, out var pillBufferSolution))
+                return;
+
+            var container = _itemSlotsSystem.GetItemOrNull(owner, SharedChemMaster.InputSlotName);
```
keep it named inputContainer to avoid conflicts

deltanedas commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -94,35 +112,34 @@ private void OnSetPillTypeMessage(Entity<ChemMasterComponent> chemMaster, ref Ch
-                !_solutionContainerSystem.TryGetSolution(chemMaster.Owner, SharedChemMaster.BufferSolutionName, out _, out var bufferSolution))
-            {
+                !_solutionContainerSystem.TryGetSolution(chemMaster.Owner, SharedChemMaster.BufferSolutionName, out _, out var bufferSolution) ||
+                !_solutionContainerSystem.TryGetSolution(chemMaster.Owner, SharedChemMaster.PillBufferSolutionName, out _, out var pillBufferSolution))
```
.

deltanedas commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -94,35 +112,34 @@ private void OnSetPillTypeMessage(Entity<ChemMasterComponent> chemMaster, ref Ch
-                return;
-
-            TransferReagents(chemMaster, message.ReagentId, message.Amount.GetFixedPoint(), message.FromBuffer);
+            TransferReagents(chemMaster, message.ReagentId, chemMaster.Comp.TransferringAmount, message.FromBuffer, message.IsOutput);
```
comment

deltanedas commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -94,35 +112,34 @@ private void OnSetPillTypeMessage(Entity<ChemMasterComponent> chemMaster, ref Ch
         }
 
-        private void TransferReagents(Entity<ChemMasterComponent> chemMaster, ReagentId id, FixedPoint2 amount, bool fromBuffer)
+        private void TransferReagents(Entity<ChemMasterComponent> chemMaster, ReagentId id, FixedPoint2 amount, bool fromBuffer, bool isOutput)
```
.

deltanedas commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -94,35 +112,34 @@ private void OnSetPillTypeMessage(Entity<ChemMasterComponent> chemMaster, ref Ch
                 amount = FixedPoint2.Min(amount, containerSolution.AvailableVolume);
-                amount = bufferSolution.RemoveReagent(id, amount, preserveOrder: true);
-                _solutionContainerSystem.TryAddReagent(containerSoln.Value, id, amount, out var _);
+                var solution = isOutput ? pillBufferSolution : bufferSolution;
```
assign this before the if statement and comment

deltanedas commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -94,35 +112,34 @@ private void OnSetPillTypeMessage(Entity<ChemMasterComponent> chemMaster, ref Ch
+                var solution = isOutput ? pillBufferSolution : bufferSolution;
+
+                amount = solution.RemoveReagent(id, amount, preserveOrder: true);
+                _solutionContainerSystem.TryAddReagent(containerSoln.Value, id, amount, out _);
```
comment

deltanedas commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -94,35 +112,34 @@ private void OnSetPillTypeMessage(Entity<ChemMasterComponent> chemMaster, ref Ch
-                bufferSolution.AddReagent(id, amount);
+
+                var solution = isOutput ? pillBufferSolution : bufferSolution;
+                solution.AddReagent(id, amount);
```
.

deltanedas commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -132,6 +149,14 @@ private void OnCreatePillsMessage(Entity<ChemMasterComponent> chemMaster, ref Ch
+                var canister = _entityManager.SpawnEntity(PillCanisterPrototypeId, Transform(chemMaster.Owner).Coordinates);
+                _itemSlotsSystem.TryInsert(chemMaster.Owner, SharedChemMaster.OutputSlotName, canister, null);
+                maybeContainer = canister;
+            }
```
comments around this block, use `Spawn` instead of `_entityManager.SpawnEntity`

deltanedas commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -162,9 +187,14 @@ private void OnCreatePillsMessage(Entity<ChemMasterComponent> chemMaster, ref Ch
+                    message.Dosage);
+
+                if (!hasItemSolution || itemSolution is null)
+                    continue;
```
untouch

deltanedas commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -173,7 +203,9 @@ private void OnCreatePillsMessage(Entity<ChemMasterComponent> chemMaster, ref Ch
-                _adminLogger.Add(LogType.Action, LogImpact.Low,
+                _adminLogger.Add(
+                    LogType.Action,
+                    LogImpact.Low,
```
.

deltanedas commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -185,11 +217,17 @@ private void OnOutputToBottleMessage(Entity<ChemMasterComponent> chemMaster, ref
+                var canister = _entityManager.SpawnEntity(PillCanisterPrototypeId, Transform(chemMaster.Owner).Coordinates);
+                _itemSlotsSystem.TryInsert(chemMaster.Owner, SharedChemMaster.OutputSlotName, canister, null);
+                maybeContainer = canister;
+            }
```
comments and Spawn

deltanedas commented on Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs:
```diff
@@ -263,42 +299,22 @@ private void ClickSound(Entity<ChemMasterComponent> chemMaster)
         }
 
-        private ContainerInfo? BuildOutputContainerInfo(EntityUid? container)
-        {
-            if (container is not { Valid: true })
-                return null;
-
-            var name = Name(container.Value);
-            {
-                if (_solutionContainerSystem.TryGetSolution(
-                        container.Value, SharedChemMaster.BottleSolutionName, out _, out var solution))
-                {
-                    return BuildContainerInfo(name, solution);
-                }
-            }
-
-            if (!TryComp(container, out StorageComponent? storage))
-                return null;
-
-            var pills = storage.Container.ContainedEntities.Select((Func<EntityUid, (string, FixedPoint2 quantity)>) (pill =>
+        private static ContainerInfo BuildContainerInfo(string name, Solution solution) =>
+            new(name, solution.Volume, solution.MaxVolume)
             {
-                _solutionContainerSystem.TryGetSolution(pill, SharedChemMaster.PillSolutionName, out _, out var solution);
-                var quantity = solution?.Volume ?? FixedPoint2.Zero;
-                return (Name(pill), quantity);
-            })).ToList();
-
-            return new ContainerInfo(name, _storageSystem.GetCumulativeItemAreas((container.Value, storage)), storage.Grid.GetArea())
-            {
-                Entities = pills
+                Reagents = solution.Contents
             };
```
bruh

sleepyyapril replied:
it no longer exists c:

deltanedas commented on Content.Shared/Chemistry/SharedChemMaster.cs:
```diff
@@ -11,6 +11,7 @@ public sealed class SharedChemMaster
     {
         public const uint PillTypes = 20;
         public const string BufferSolutionName = "buffer";
+        public const string PillBufferSolutionName = "pillBuffer";
```
comment

deltanedas commented on Content.Shared/Chemistry/SharedChemMaster.cs:
```diff
@@ -33,14 +34,16 @@ public ChemMasterSetPillTypeMessage(uint pillType)
     {
         public readonly ReagentId ReagentId;
-        public readonly ChemMasterReagentAmount Amount;
+        public readonly int Amount;
```
.

deltanedas commented on Content.Shared/Chemistry/SharedChemMaster.cs:
```diff
@@ -33,14 +34,16 @@ public ChemMasterSetPillTypeMessage(uint pillType)
+        public readonly bool IsOutput;
 
-        public ChemMasterReagentAmountButtonMessage(ReagentId reagentId, ChemMasterReagentAmount amount, bool fromBuffer)
+        public ChemMasterReagentAmountButtonMessage(ReagentId reagentId, int amount, bool fromBuffer, bool isOutput)
```
.

deltanedas commented on Content.Shared/Chemistry/SharedChemMaster.cs:
```diff
@@ -33,14 +34,16 @@ public ChemMasterSetPillTypeMessage(uint pillType)
-        public readonly ChemMasterReagentAmount Amount;
+        public readonly int Amount;
         public readonly bool FromBuffer;
+        public readonly bool IsOutput;
```
.

deltanedas commented on Content.Shared/Chemistry/SharedChemMaster.cs:
```diff
@@ -33,14 +34,16 @@ public ChemMasterSetPillTypeMessage(uint pillType)
             ReagentId = reagentId;
             Amount = amount;
             FromBuffer = fromBuffer;
+            IsOutput = isOutput;
```
.

ThataKat commented:
Reviewing, 24-48 hours

sleepyyapril committed:
review stuff

sleepyyapril committed:
Better comments?

sleepyyapril committed:
Merge branch 'master' into port-chemmaster-plus

deltanedas commented on Content.Client/Chemistry/UI/ChemMasterBoundUserInterface.cs:
```diff
@@ -34,8 +34,6 @@ protected override void Open()
             // Setup static button actions.
             _window.InputEjectButton.OnPressed += _ => SendMessage(
                 new ItemSlotButtonPressedEvent(SharedChemMaster.InputSlotName));
-            _window.OutputEjectButton.OnPressed += _ => SendMessage(
-                new ItemSlotButtonPressedEvent(SharedChemMaster.OutputSlotName));
```
DeltaV comment

deltanedas reviewed (changes requested):
all these comments should be `// DeltaV` not Delta-v

sleepyyapril committed:
Delta-v is kill

sleepyyapril committed:
Merge branch 'port-chemmaster-plus' of https://github.com/sleepyyapril/Delta-v into port-chemmaster-plus

sleepyyapril committed:
comment here?

deltanedas commented on Resources/Locale/en-US/chemistry/components/chem-master-component.ftl:
move all of these new strings to _DV so theres no changes to this file

deltanedas commented on Resources/Prototypes/Entities/Structures/Machines/chem_master.yml:
```diff
@@ -86,6 +86,7 @@
   - type: SolutionContainerManager
     solutions:
       buffer: {}
+      pillBuffer: {}
```
comment

deltanedas commented on Content.Shared/Chemistry/SharedChemMaster.cs:
```diff
@@ -131,36 +124,41 @@ public ContainerInfo(string displayName, FixedPoint2 currentVolume, FixedPoint2
+        public readonly bool UpdateLabel = updateLabel;
+
+        public readonly int SortMethod = sortMethod;
+        public readonly int TransferringAmount = transferringAmount;
```
undo the reformatting this changes every single line

deltanedas commented on Content.Shared/Chemistry/SharedChemMaster.cs:
```diff
@@ -72,26 +75,16 @@ public ChemMasterOutputToBottleMessage(uint dosage, string label)
-            else
-                return FixedPoint2.New((int)amount);
-        }
+        public readonly int TransferringAmount = transferringAmount;
```
comment

sleepyyapril committed:
Merge branch 'master' of https://github.com/DeltaV-Station/Delta-v into port-chemmaster-plus

sleepyyapril committed:
Bro made me comment

sleepyyapril committed:
fix deltav

sleepyyapril committed:
fully untouch ftl

sleepyyapril committed:
touch less

sleepyyapril committed:
touch even less

sleepyyapril committed:
even less

sleepyyapril committed:
less

sleepyyapril committed:
comment more!

sleepyyapril committed:
Merge branch 'master' into port-chemmaster-plus

deltanedas commented on Resources/Locale/en-US/chemistry/components/chem-master-component.ftl:
```diff
@@ -10,7 +10,7 @@ chem-master-bound-user-interface-title = ChemMaster 4000
 chem-master-window-input-tab = Input
 chem-master-window-output-tab = Output
-chem-master-window-container-label = Container
+chem-master-window-container-label =   Container
```
untouch

deltanedas commented on Resources/Locale/en-US/_DV/chemistry/components/chem-master-component.ftl:
```diff
@@ -0,0 +1,16 @@
+chem-master-window-transferring-label =    Transferring: [color={$color}]{$quantity}[/color]
+chem-master-window-transferring-default-label =    Transferring: [color=#ffffff]50[/color]
+chem-master-window-reagent-move-button = Move
+chem-master-window-discard-button = Discard
```
file needs trailing newline

deltanedas commented on Content.Shared/Chemistry/SharedChemMaster.cs:
```diff
@@ -148,18 +140,40 @@ public sealed class ChemMasterBoundUserInterfaceState : BoundUserInterfaceState
+
+        public readonly FixedPoint2? PillBufferCurrentVolume; // DeltaV
+
+        public ChemMasterBoundUserInterfaceState(  // DeltaV
+            ContainerInfo? inputContainerInfo,
+            IReadOnlyList<ReagentQuantity> bufferReagents,
+            IReadOnlyList<ReagentQuantity> pillBufferReagents,
+            FixedPoint2 bufferCurrentVolume,
+            FixedPoint2 pillBufferCurrentVolume,
+            uint selectedPillType,
+            uint pillDosageLimit,
+            bool updateLabel,
+            int sortMethod,
+            int transferringAmount)
```
this still reformats the constructor, keep it how it was and move all deltav fields to the end

ThataKat commented:
Working this one out with the author through discord channels, review still in progress

Lyndomen reviewed (approved):
Direction Approved, but requesting a do not merge till April gives the go ahead that all the requested features have been ported

deltanedas commented:
fell off and the pills stuff exists now which is the main thing
